### PR TITLE
oded/act-468

### PR DESCRIPTION
**Pull request by odeds**

### CODE REVIEW

File: dist/znk.infra.js [hunk 1]
```diff
@@ -645,6 +645,9 @@
 
                         var exerciseResultPath = _getExerciseResultPath(newGuid);
                         var initResult = _getInitExerciseResult(exerciseTypeId,exerciseId,newGuid);
+                        if(examSectionsNum) {
```
Comment (igormas) [inline dist/znk.infra.js]: Why we need this condition ? cab examId property just be set?

File: dist/znk.infra.js [hunk 2]
```diff
@@ -747,6 +755,29 @@
 
             }
 
+            function _saveExamWhenAllSectionsCompleted(exerciseNewStatus, exerciseResult, exercisesStatusData) {
```
Comment (igormas) [inline dist/znk.infra.js]: Bad implementation, the save action should be atomic !!!

File: dist/znk.infra.js [hunk 3]
```diff
@@ -747,6 +755,29 @@
...
+                        if(examResults.examSectionsNum && sectionResultsToArr.length === +examResults.examSectionsNum) {
+                            for(var i = 0, ii = sectionResultsToArr.length; i < ii; i++) {
+                                if(exercisesStatusDataPerExerciseTypeId[sectionResultsToArr[i]].status !== ExerciseStatusEnum.COMPLETED.enum) {
+                                    isExamComplete = false;
+                                    break;
+                                }
+                            }
+                            if(isExamComplete) {
+                                examResults.isComplete = true;
+                                examResults.endedTime = Date.now();
```
Comment (igormas) [inline dist/znk.infra.js]: Firebase time stamp should used instead.

File: dist/znk.infra.js [hunk 4]
```diff
@@ -737,6 +743,8 @@
                         ExerciseStatusEnum.COMPLETED.enum : ExerciseStatusEnum.ACTIVE.enum;
                     exercisesStatusData[exerciseResult.exerciseTypeId][exerciseResult.exerciseId] = new ExerciseStatus(exerciseNewStatus);
 
+                    _saveExamWhenAllSectionsCompleted(exerciseNewStatus, exerciseResult, exercisesStatusData);
```
Comment (igormas) [inline dist/znk.infra.js]: What about the case in which this is not section result ?

File: test/spec/exerciseResult/exerciseResult.spec.js [hunk 5]
```diff
@@ -353,4 +353,55 @@ describe('testing service "ExerciseResult":', function () {
             expect(exerciseResult.avgTimePerQuestion).toBe(expectedResult);
         });
     });
+
+    describe('test save exam isCompleted when all sections are completed',function(){
+
+        it('when getExerciseResult called with examSectionsNum as the forth param, then it should add examSectionsNum and examId on exerciseResult', function(){
```
Comment (igormas) [inline test/spec/exerciseResult/exerciseResult.spec.js]: Exam id should be always added to new section result!!!

File: test/spec/exerciseResult/exerciseResult.spec.js [hunk 6]
```diff
@@ -353,4 +353,55 @@ describe('testing service "ExerciseResult":', function () {
...
+            var examId = 45;
+            var exerciseResult = actions.getExerciseResult(ExerciseTypeEnum.SECTION.enum, exerciseId, examId, examSectionsNum);
+            exerciseResult.$save();
+            $rootScope.$digest();
+            var examResult = actions.getExamResult(examId);
+            expect(exerciseResult.examId).toEqual(examId);
+            expect(examResult.examSectionsNum).toEqual(examSectionsNum);
+        });
+
+        it('when _saveExamWhenAllSectionsCompleted with exerciseNewStatus that is not 2, then getExamResult shouldn\'t been called inside the _saveExamWhenAllSectionsCompleted, total calls 1', function(){
```
Comment (igormas) [inline test/spec/exerciseResult/exerciseResult.spec.js]: The required tests are: 
* Saving not completed section result -> exam is completed result should not be true.
* Saving completed section result while other exam sections are were not started-> exam is completed result should not be true.
* Saving completed section result while other exam sections are active but not completed -> exam is completed result should not be true.
* Saving completed section result while other exam sections are also completed -> exam is completed result should be true.

File: dist/znk.infra.js (hunk 1 above)
Comment (odeds) [inline dist/znk.infra.js]: ye ok i guess so.

File: dist/znk.infra.js (hunk 2 above)
Comment (odeds) [inline dist/znk.infra.js]: changed it

File: dist/znk.infra.js (hunk 3 above)
Comment (odeds) [inline dist/znk.infra.js]: done

File: test/spec/exerciseResult/exerciseResult.spec.js (hunk 5 above)
Comment (odeds) [inline test/spec/exerciseResult/exerciseResult.spec.js]: changed it

File: test/spec/exerciseResult/exerciseResult.spec.js (hunk 6 above)
Comment (odeds) [inline test/spec/exerciseResult/exerciseResult.spec.js]: done

File: dist/znk.infra.js (hunk 4 above)
Comment (odeds) [inline dist/znk.infra.js]: changed it

File: test/spec/exerciseResult/exerciseResult.spec.js [hunk 7]
```diff
@@ -353,4 +353,98 @@ describe('testing service "ExerciseResult":', function () {
...
+            examResult = actions.getExamResult(examId);
+            expect(examResult.isComplete).toEqual(false);
+        });
+
+        it('when section is complete, and other section are also completed, then examResult.isComplete = true', function(){
+            var exerciseId = 4;
+            var examSectionsNum = 5;
+            var examId = 45;
+            var sectionResultsObj = {
+                1: '111',
```
Comment (assafshp) [inline test/spec/exerciseResult/exerciseResult.spec.js]: what is this ? I'm not familiar with such struct for section results object ?

File: test/spec/exerciseResult/exerciseResult.spec.js [hunk 8]
```diff
@@ -353,4 +353,98 @@ describe('testing service "ExerciseResult":', function () {
...
+            $rootScope.$digest();
+            var examResult = actions.getExamResult(examId);
+            expect(examResult.isComplete).toEqual(false);
+        });
+
+        it('when section is complete, and other section are started but not completed, then examResult.isComplete = false', function(){
+            var exerciseId = 4;
+            var examSectionsNum = 5;
+            var examId = 45;
+            var sectionResultsObj = {
```
Comment (assafshp) [inline test/spec/exerciseResult/exerciseResult.spec.js]: what is this ? I'm not familiar with such struct for section results object ?

File: test/spec/exerciseResult/exerciseResult.spec.js [hunk 9]
```diff
@@ -353,4 +353,98 @@ describe('testing service "ExerciseResult":', function () {
...
+                };
+            }
+            var examResult = actions.getExamResult(examId);
+            examResult.sectionResults = sectionResultsObj;
+            examResult.$save();
+            $rootScope.$digest();
+            var exerciseResult = actions.getExerciseResult(ExerciseTypeEnum.SECTION.enum, exerciseId, examId, examSectionsNum);
+            exerciseResult.isComplete = true;
+            exerciseResult.$save();
+            $rootScope.$digest();
```
Comment (assafshp) [inline test/spec/exerciseResult/exerciseResult.spec.js]: if it is atomic why can't you use only one digest?

File: test/spec/exerciseResult/exerciseResult.spec.js (hunk 7 above)
Comment (odeds) [inline test/spec/exerciseResult/exerciseResult.spec.js]: the keys are like section id and ref section guid, we care only on the number of section results

File: test/spec/exerciseResult/exerciseResult.spec.js (hunk 8 above)
Comment (odeds) [inline test/spec/exerciseResult/exerciseResult.spec.js]: the keys are like section id and ref section guid, we care only on the number of section results

File: test/spec/exerciseResult/exerciseResult.spec.js (hunk 9 above)
Comment (odeds) [inline test/spec/exerciseResult/exerciseResult.spec.js]: removed the extra digest

**Comment by assafshp** [comment]
conflict

**Pull request merged**